File: exercises/exercise_2/hello-world.c
// gcc 4.7.2 +
// gcc -std=gnu99 -Wall -g -o helloworld_c helloworld_c.c -lpthread

#include <pthread.h>
#include <stdio.h>

pthread_mutex_t lock;

int count = 0;
// Note the return type: void*
void* someThreadFunction(){
	pthread_mutex_lock(&lock);

    int i;
    for(i = 0; i < 1000000; ++i){
    	count++;
    }

    pthread_mutex_unlock(&lock);
    return NULL;
}

void* someThreadFunction2(){
    pthread_mutex_lock(&lock);

    int i;
    for(i = 0; i < 1000000; ++i){
    	count--;
    }

    pthread_mutex_unlock(&lock);
    return NULL;
}


int main(){
    pthread_t someThread;
    pthread_create(&someThread, NULL, someThreadFunction, NULL);
    // Arguments to a thread would be passed here ---------^
    pthread_t someThread2;
    pthread_create(&someThread2, NULL, someThreadFunction2, NULL);
    
    pthread_join(someThread, NULL);
    pthread_join(someThread2, NULL);

    printf("count equals: %d",count);
    return 0;
    
}

File: exercises/exercise_2/hello-world.py

# Python 3.3.3 and 2.7.6
# python helloworld_python.py

import threading

count = 0
lock = threading.Lock()

def someThreadFunction():
    global count
    lock.acquire()
    print("Hello from a thread!")
    for i in xrange(1000000):
    	count+=1
    lock.release()

def someThreadFunction2():
    global count
    lock.acquire()
    print("Hello from a thread!")
    for i in xrange(1000000):
        count-=1
    lock.release()
# Potentially useful thing:
#   In Python you "import" a global variable, instead of "export"ing it when you declare it
#   (This is probably an effort to make you feel bad about typing the word "global")
    


def main():
    someThread = threading.Thread(target = someThreadFunction, args = (),)
    someThread2 = threading.Thread(target = someThreadFunction2, args = (),)

    someThread.start()
    someThread2.start()

    someThread.join()
    someThread2.join()
    print("Count equals: ", count)


main()